app/controllers/auth_controller.rb

```diff
@@ -30,4 +36,32 @@ def auth_with_okta
     render_bad_request e.message
     nil
   end
+
+  def auth_with_facebook
+    return unless /^[0-9a-zA-Z]/.match?(data_attrs[:facebook_access_token])
+    url = "https://graph.facebook.com/debug_token?input_token=#{data_attrs[:facebook_access_token]}&access_token=#{ENV.fetch("FACEBOOK_APP_ID")}|#{ENV.fetch("FACEBOOK_APP_SECRET")}"
+
+    begin
+      data = JSON.parse(Net::HTTP.get(URI(url)))
+    rescue Net::HTTPBadRequest => e
+      render_bad_request(e.to_s) and return
+    end
+
+    (render_bad_request(data["data"]["error"]) and return) if data["data"] && data["data"]["error"]
+    (render_bad_request("facebook token is not valid") and return) unless data["data"] && data["data"]["is_valid"] && data["data"]["user_id"]
+
+    user_id = data["data"]["user_id"]
+    url = "https://graph.facebook.com/#{user_id}?fields=email,id,first_name,last_name,short_name&access_token=#{data_attrs[:facebook_access_token]}"
+    begin
+      data = JSON.parse(Net::HTTP.get(URI(url)))
+    rescue Net::HTTPBadRequest
+      render_bad_request("could not extract facebook fields required") and return
+    end
+
+    (render_bad_request(data["data"]["error"]) and return) if data["data"] && data["data"]["error"]
+
+    user = User.where(email: data["email"]).first_or_create
+    user.update!(facebook_user_id: data["id"], first_name: data["first_name"], last_name: data["last_name"], short_name: data["short_name"])
```

Comment:
I would like to default the user lookup to prefer looking up the user by the `facebook_user_id` first. I think I'm fine with a fallback to looking the user up by email if the user_id doesn't match.

Reply:
side note, is `email` listed as needing to be unique in the `User` model? we might need to relax that if so because I could come up with a scenario where 2 accounts would have the same email address based on a user updating email addresses/creating a new account.

Reply:
falling back to email address lookup could have other implications for how data is linked to individual accounts. so just do `facebook_user_id` matching for now.

Reply:
Scenario 1:

1. `fbUser1` logs in with `email1`
   - This create `mcUser1`
2. `fbUser1` changes the email on their facebook account to `email2`
3. `fbUser2` creates/updates their account to use `email1` as their email
4. `fbUser2` logs in with `email1`
   - The current algorithm will find `mcUser1` for `fbUser2` and return that as the logged in user

Reply:
Ah, right, ok. Makes sense.